### ISSUE TITLE
Change the signature of the functions defined in the `Geocoder` interface

### ISSUE THREAD

**Issue by manolovl** [codingsince1985/geo-golang#29]
More often one would like to have an address, returned by reverse geocoding service, as a structure, with separate fields for all the parts an address consists of. Returning a formatted address doesn't cut it most of the times. Also (as a personal opinion of mine and engineers in my team) I find it nice having homogeneous definitions - `Geocode` returns the `Location` structure, so why not have a similar signature for `ReverseGeocode`.
 Hence I would like to propose (and implement it all over) the following change in the `Geocoder` interface definitions: 

```Go
type Geocoder interface {
	Geocode(address string) (*Location, error)
	ReverseGeocode(lat, lng float64) (*Address, error)
}
```
with `Address` being defined in the package `geo` in a similar manner to the `Location` one

```Go
// Address is the structured representation of an address, including its flat representation
type Address struct {
	FormattedAddress string
	Street           string
	HouseNumber      string
	Suburb           string
	Postcode         string
	State            string
	StateDistrict    string
	County           string
	Country          string
	CountryCode      string
	City             string
}
```

My reasoning for the further proposed changes:
- Return pointer to the defined structure, so that the caller can check for `nil`; checking for zero values to deduce `ErrNoResult` may often go wrong (further no result being returned shouldn't be an error)
- Add second return value of type `error` makes is easier to catch errors and propagate them back to the caller, next to a `nil`, who then should decide how to react
- The (uniform) address string, currently returned by `Address()` would be accessible through the `FormattedAddress` field

**Comment by codingsince1985** [comment]
Sounds good.

**Comment by manolovl** [comment]
Just a quick question. Is there a specific reason why the tests for the specific providers/clients are in different packages? Based on what I've seen, I usually make the tests part of the same package in which the code under test is.

**Comment by codingsince1985** [comment]
I remember It's done this way simply because test cases have the same visibility (of the library) as real life library users do. Definitely not something can't be changed.

**Comment by manolovl** [comment]
I see, thanks for the clarification.

**Comment by codingsince1985** [comment]
Tip2 of [this](https://medium.com/@povilasve/go-advanced-tips-tricks-a872503ac859) ;-)

**Comment by manolovl** [comment]
Yes, such a good article! :)
BTW, I'm almost done with this here, just some clean-up (gofmt, goimports). last check on tests and should be ready with a PR (finally!).

**Comment by manolovl** [comment]
Hi @codingsince1985,

anything against renaming `geocoder_test.go` to something like `geocoder_examples.go`, so it won't get executed as a test?

Cheers,
L

**Comment by manolovl** [comment]
Update: sorry, I meant the `geocoder_test.go` from the root dir, of course :)

**Comment by codingsince1985** [comment]
No.

**Comment by manolovl** [comment]
Resolved by [PR#34](https://github.com/codingsince1985/geo-golang/pull/34)
